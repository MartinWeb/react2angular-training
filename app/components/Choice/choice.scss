.result {
    margin-bottom: 50px;
    margin-top: 20px;

    form {
        input[type=text] {
            width: 300px;
            height: 32px;
            font-size: 14px;
            margin-right: 10px;
        }

        input[type=submit] {
            padding: 10px;
            border-radius: 0;
            background-color: rgb(241, 90, 36);
            color: white;
            border: 1px solid rgb(241, 90, 36);

            &:hover {
                cursor: pointer;
            }
        }
    }
}

.choice {
    h4 {
        text-decoration: underline;
    }
    
    .choice_radio {
        background-color: dodgerblue;
        color: white;
        padding: 10px;
        display: inline;
        margin-right: 10px;

        &:hover, & > input[type=radio], & > label {
            cursor: pointer;
        }
    }
}