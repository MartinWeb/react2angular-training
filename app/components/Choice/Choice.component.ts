import angular, { IScope } from "angular";
import { ApiService } from "services/api.service";
import template from './choice-template.html';
import './choice.scss';

export class ChoiceController implements angular.IController {
    public choice: string;
    public search: string;
    public joke: ChuckNorris.Joke = null;
    public jokes: ChuckNorris.Joke[] = null;
    public categories: string[] = null;

    constructor(private readonly apiService: ApiService, private readonly $scope: IScope) {}

    public onSelectChoice = async () => {
        if(this.choice === 'random') {
            this.joke = await this.apiService.getJokesRandomly();
            this.$scope.$apply();
        } 
        
        if(this.choice === 'categories') {
            this.categories = await this.apiService.getJokesCategories();
            this.$scope.$apply();
        }
    }

    public onSearch = async () => {
        this.jokes = await this.apiService.getJokeBySearch(this.search);

        this.$scope.$apply();
    }
}

ChoiceController.$inject = ['apiService', '$scope'];

export const ChoiceComponent: angular.IComponentOptions = {
    template,
    controller: ChoiceController
};