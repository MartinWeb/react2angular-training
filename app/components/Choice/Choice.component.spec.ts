import { IApiService } from "services/api.service";
import { ChoiceComponent, ChoiceController } from "./Choice.component";

describe('ChoiceComponent', () => {
    test('should have valid snapshot', () => {
        expect(ChoiceComponent.template).toMatchSnapshot();
    });

    describe('ChoiceController', () => {
        let target: ChoiceController;
        let $scope: angular.IScope;
        let apiService: IApiService;
        const resultJoke = { value: 'joke' } as ChuckNorris.Joke;
        const resultCategories = ['animal', 'food'];
        const resultJokes = [{ value: 'joke' } as ChuckNorris.Joke, { value: 'joke 2' } as ChuckNorris.Joke];

        beforeEach(() => {
            $scope = ({
                $apply: jest.fn(),
            } as unknown) as angular.IScope;

            apiService = ({
                getJokesRandomly: jest.fn().mockResolvedValue(resultJoke),
                getJokesCategories: jest.fn().mockResolvedValue(resultCategories),
                getJokeBySearch: jest.fn().mockResolvedValue(resultJokes)
            } as unknown) as IApiService;

            target = new ChoiceController(apiService, $scope);
        });

        it('should have correct injection', () => {
            expect(ChoiceController.$inject).toEqual(['apiService', '$scope']);
        });

        describe('onSearch', () => {
            beforeEach(async () => {
                target.search = 'christmas';
                await target.onSearch();
            });

            it('should call apiService', async () => {
                expect(apiService.getJokeBySearch).toHaveBeenCalledWith('christmas');
            });

            it('should return result', async () => {
                expect(target.jokes).toEqual(resultJokes);
            });

            it('should call $apply', async () => {
                expect($scope.$apply).toHaveBeenCalled();
            });
        });

        describe('onSelectChoice', () => {
            describe('choice random', () => {
                beforeEach(async () => {
                    target.choice = 'random';
                    await target.onSelectChoice();
                });
                
                it('should call apiService', async () => {
                    await expect(apiService.getJokesRandomly).toHaveBeenCalled();
                });
    
                it('should return result', async () => {
                    await expect(target.joke).toEqual(resultJoke);
                });
    
                it('should call $apply', async () => {
                    await expect($scope.$apply).toHaveBeenCalled();
                });
            });

            describe('choice categories', () => {
                beforeEach(async () => {
                    target.choice = 'categories';
                    await target.onSelectChoice();
                });

                it('should call apiService', async () => {
                    await expect(apiService.getJokesCategories).toHaveBeenCalled();
                });
    
                it('should return result', async () => {
                    await expect(target.categories).toEqual(resultCategories);
                });
    
                it('should call $apply', async () => {
                    await expect($scope.$apply).toHaveBeenCalled();
                });
            });
        });
    });
});