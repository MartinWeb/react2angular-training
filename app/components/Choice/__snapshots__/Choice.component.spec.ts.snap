// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChoiceComponent should have valid snapshot 1`] = `
"<div class="choice">
    <h4>Chuck Norris is full of knowledge, but what would you like to do now ? Which way do you want to interrogate Chuck ?</h4>

    <form>
        <label for="random" class="choice_radio">
            <input type="radio" id="random" name="choice" value="random" ng-model="$ctrl.choice" ng-change="$ctrl.onSelectChoice()">
            Randomly
        </label>
        
        <label for="categories" class="choice_radio">
            <input type="radio" id="categories" name="choice" value="categories" ng-model="$ctrl.choice" ng-change="$ctrl.onSelectChoice()">
            Show me some categories
        </label>

        <label for="search" class="choice_radio">
            <input type="radio" id="search" name="choice" value="search" ng-model="$ctrl.choice" ng-change="$ctrl.onSelectChoice()">
            I want to search a particular word
        </label>
    </form>

    <div class="result">
        <div ng-if="$ctrl.choice === 'search'">
            <p>Type the word you want Chuck Norris to tell you about :</p>
            <form>
                <input type="text" ng-model="$ctrl.search" />
                <input type="submit" value="Search" ng-click="$ctrl.onSearch()" />
            </form>
    
            <div ng-if="$ctrl.search !== null && $ctrl.jokes !== null">
                <search-result-component jokes="$ctrl.jokes"></search-result-component>
            </div> 
        </div>
    
        <div ng-if="$ctrl.choice === 'random' && $ctrl.joke !== null">
            <random-result-component joke="$ctrl.joke" on-reload="$ctrl.onSelectChoice"></random-result-component>
        </div> 
    
        <div ng-if="$ctrl.choice === 'categories' && $ctrl.categories !== null">
            <categories-result-component categories="$ctrl.categories"></categories-result-component>
        </div> 
    </div>
</div>"
`;
